src/sections/upload-dataset-files/UploadDatasetFiles.tsx

```diff
@@ -42,15 +46,36 @@ export const UploadDatasetFiles = ({ fileRepository: fileRepository }: UploadDat
     })
   }
 
-  const fileUploadFinished = (file: File) => {
-    const key = FileUploadTools.key(file)
+  const fileUploadFailed = (file: File) => {
     setUploadingToCancelMap((x) => {
-      x.delete(key)
+      x.delete(FileUploadTools.key(file))
       return x
     })
     releaseSemaphore(file)
   }
 
+  const fileUploadFinished = (file: File) => {
+    const hash = md5.create()
```

Comment:
Just to understand the code better, why do we need to store the checksum of the uploaded file?
And notice that in the .finally block, the code may succeed or may fail, the finally block runs when the promise is settled but runs on both succeed or fail of the promise

Reply:
Uploading files to a dataset is a two-step process: first upload all files to s3, then add all uploaded files (this call requires the checksum) to the dataset. The checksum is needed when we call "add uploaded files" use-case, but it is calculated immediately after uploading a specific file to s3 ("upload file" use-case), while the semaphore is still taken. The semaphore makes sure that we can upload/calculate checksum only for a certain number of files concurrently.

That said, reading a file should not fail while calculating a checksum, unless you delete that file after it was uploaded to s3 on your computer, but before the checksum calculation finished. In that case, the value for checksum is not set and addUploadedFiles will fail. Regardless of the success of checksum calculations, we need to prevent the resource leak (by keeping the reference to the cancel function that can no longer be used, as the upload to s3 finished) and keeping the semaphore (that would freeze the application, as the upload to s3 is queued on that semaphore). The `finally` block release these resources.

Reply:
Thanks for the detailed explanation @ErykKul 👏🏼 👏🏼